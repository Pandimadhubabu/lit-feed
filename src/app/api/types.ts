export type ErrorData = Record<string, unknown>;

export type HttpResponse = {
  status: 200 | 201 | 204 | 400 | 401 | 403 | 404 | 500;
  message: string;
  data?: unknown;
};

export type NextRequestWithParams = {
  body?: Record<string, unknown>;
  query?: Record<string, unknown>;
  params?: Record<string, unknown>;
  url?: string;
};

export interface OauthClaims {
  nickname: string;
  name: string;
  picture: string;
  updated_at: string;
  email: string;
  email_verified: string;
  sub: string;
  sid: string;
}
