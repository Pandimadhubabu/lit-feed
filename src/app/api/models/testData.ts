import { Article, Feed, User } from "@/types";
import { OauthClaims } from "../types";

export const testFeeds: Feed[] = [
  {
    href: "https://www.reddit.com/r/programming/.rss",
    id: "6520691d42b827145292f2f1",
    name: "programming",
    unread: 0,
  },
  {
    href: "https://www.reddit.com/r/javascript/.rss",
    id: "6520691d42b827145292f2f2",
    name: "javascript",
    unread: 0,
  },
];

export const testArticles: Article[] = [
  {
    feedId: "6520691d42b827145292f2f1",
    id: "245f23984f233d32b233f2f1",
    title: "test",
    date: new Date().toISOString(),
    duration: "5 minutes",
    feedName: "programming",
    href: "https://www.reddit.com/r/programming/comments/1ykrzo/why_i_use_vim/",
    isRead: false,
    isSaved: false,
    summary: "test",
  },
  {
    feedId: "6520691d42b827145292f2f1",
    id: "245f23984f233d32b233f2f2",
    title: "test",
    date: new Date().toISOString(),
    duration: "5 minutes",
    feedName: "javascript",
    href: "https://www.reddit.com/r/javascript/comments/1ykrzo/why_i_use_vim/",
    isRead: false,
    isSaved: false,
    summary: "test",
  },
];

export const testUser: User = {
  email: "test@example.com",
  name: "Test User",
  picture: "https://example.com/test.jpg",
  id: "245f23984f234d32b233f2f2",
  oauthId: "OAUTH_ID",
  isEmailVerified: true,
  nickname: "test",
  updatedAt: new Date("2020-01-01T00:00:00.000Z"),
};

export const testClaims: OauthClaims = {
  email: testUser.email,
  email_verified: "true",
  name: testUser.name,
  nickname: testUser.nickname,
  picture: testUser.picture,
  sid: "SESSION_ID",
  sub: testUser.oauthId,
  updated_at: testUser.updatedAt.toISOString(),
};
