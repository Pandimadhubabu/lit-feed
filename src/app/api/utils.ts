export function omit(obj: Record<never, unknown>, keys: string[]) {
  const newObj = { ...obj };
  keys.forEach((key) => delete (newObj as never)[key]);
  return newObj;
}
