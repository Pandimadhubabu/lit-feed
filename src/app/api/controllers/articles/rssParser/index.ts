import { Article } from "@/types";
import { XMLParser } from "fast-xml-parser";
import * as logger from "../../../logger";
import { ServerError } from "@/app/api/errors";

export async function getParsedArticles(url: string): Promise<ParsedArticle[]> {
  logger.debug({ url }, "Fetching articles from feed");
  const rssResponse = await fetch(url);

  const rssText = await rssResponse.text();
  logger.debug({ rssText }, "Fetched articles from feed");

  const rss = new XMLParser().parse(rssText);

  const rssItems = getRSSItems(rss);

  return rssToArticles(rssItems);
}

export function rssToArticles(rss: RSSItem[]): ParsedArticle[] {
  const articles: ParsedArticle[] = rss.map((item: RSSItem) => ({
    title: getTitle(item),
    summary: getDescription(item),
    href: getLink(item),
    date: getDate(item),
    image: getImage(item),
  }));

  logger.debug({ articles }, "Parsed articles from feed");

  return articles;
}

export function getTitle(item: RSSItem) {
  const title = getSimilarField(item, "title");

  if (!title) {
    throw new ServerError("No title found", { item });
  }
  return title;
}

export function getDescription(item: RSSItem) {
  const description = getSimilarField(item, "description");

  if (!description) {
    throw new ServerError("No description found", { item });
  }
  return description;
}

export function getLink(item: RSSItem) {
  const link = getSimilarField(item, "link");

  if (!link) {
    throw new ServerError("No link found", { item });
  }
  return link;
}

export function getImage(item: RSSItem) {
  return getSimilarField(item, "image");
}

export function getDate(item: RSSItem) {
  if (item.pubDate) {
    return item.pubDate;
  }
  return getSimilarField(item, "date");
}

export function getSimilarField(item: RSSItem, field: string) {
  if (item[field]) {
    return item[field];
  }
  const keys = Object.keys(item);

  for (const key of keys) {
    if (key.match(new RegExp(field, "i"))) {
      return item[key];
    }
  }
}

export function getRSSItems(rssInput: Record<string, any>): RSSItem[] {
  const rss = rssInput?.rss || rssInput?.feed || rssInput?.rdf || rssInput;
  if (rss?.channel?.item) {
    return rss.channel.item;
  }
  if (rss?.item) {
    return rss.item;
  }
  if (rss?.channel?.entry) {
    return rss.channel.entry;
  }
  if (rss?.entry) {
    return rss.entry;
  }
  if (rss?.channel?.items) {
    return rss.channel.items;
  }
  if (rss?.items) {
    return rss.items;
  }
  throw new ServerError("No items found", { rss });
}

export type ParsedArticle = Partial<Article> & {
  title: string;
  summary: string;
  href: string;
};

type RSSItem = Record<string, string>;
