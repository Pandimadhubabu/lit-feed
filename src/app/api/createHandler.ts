import { UnauthorizedError } from "./errors";
import { Repository, createRepository } from "./models/Repository";
import { NextRequestWithParams } from "./types";

type MethodToDataType<U> = {
  get: U;
  getAll: U[];
  create: U;
  delete: { id: string };
  update: { id: string };
};

type MethodReturnType<
  R,
  M extends "get" | "getAll" | "create" | "delete" | "update",
> = R extends typeof Repository<infer U> ? MethodToDataType<U>[M] : never;

export function createHandlerFromRepository<
  R extends typeof Repository<unknown>,
  M extends "get" | "getAll" | "create" | "delete" | "update",
>(
  // class type of Repository
  RepositoryClass: R,
  method: M,
) {
  return async (request: NextRequestWithParams) => {
    const { user } = request;
    if (!user) {
      throw new UnauthorizedError("Unauthorized");
    }
    const repository = createRepository(RepositoryClass, user);
    const data = (await repository[method](request)) as MethodReturnType<R, M>;

    return {
      status: 200,
      data,
      message: `Performed ${method} on ${RepositoryClass.name} successfully`,
      // infer data type from the method used on the repository
    } as const;
  };
}
