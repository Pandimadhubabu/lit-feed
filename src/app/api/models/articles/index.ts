import { debug } from "@/app/api/logger";
import { Article, Feed, User } from "@/types";
import { articles, feeds, mongoToObject, objectToMongo } from "../mongo";
import { ObjectId } from "mongodb";
import { NotFoundError } from "@/app/api/errors";

class ArticlesRepository {
  constructor(private user: User) {}

  async getArticles(feedId: Feed["id"]): Promise<Article[]> {
    const articlesCollection = await articles();

    const mongoResult = await articlesCollection
      .find({
        feedId,
      })
      .toArray();

    debug({ mongoResult }, "mongoResult");

    const articlesList = mongoResult.map(mongoToObject<Article>);

    debug({ articlesList }, "articlesList");

    return articlesList;
  }
  async addArticle(originalArticle: Article, feedId: Feed["id"]) {
    const article = Object.assign({}, originalArticle);
    const feedsCollection = await feeds();

    const feed = await feedsCollection.findOne({
      _id: new ObjectId(feedId),
    });

    debug({ feed }, "feed");

    if (!feed) {
      throw new NotFoundError("No feed found", { feedId });
    }

    const articlesCollection = await articles();

    const { insertedId } = await articlesCollection.insertOne({
      ...article,
      feedId,
    });

    debug({ insertedId }, "mongoResult");

    const articleResult = mongoToObject<Article>({
      ...article,
      _id: insertedId,
    });

    debug({ articleResult }, "addArticle result");

    return articleResult;
  }
  async getArticle(id: Article["id"]): Promise<Article> {
    const articlesCollection = await articles();

    const mongoResult = await articlesCollection.findOne(
      objectToMongo({
        id,
      }),
    );

    debug({ mongoResult }, "mongoResult");

    if (!mongoResult) {
      throw new NotFoundError("No article found", { id });
    }

    const article = mongoToObject<Article>(mongoResult);

    debug({ article }, "article");

    return article;
  }
  async deleteArticle(id: Article["id"]) {
    const articlesCollection = await articles();

    const mongoResult = await articlesCollection.deleteOne({
      _id: new ObjectId(id),
    });

    debug({ mongoResult }, "mongoResult");

    if (!mongoResult.deletedCount) {
      throw new NotFoundError("No article found", { id });
    }

    return {
      id,
    };
  }
  async updateArticle(originalArticle: Article, id: Article["id"]) {
    const article = Object.assign({}, originalArticle);
    const articlesCollection = await articles();

    const mongoResult = await articlesCollection.updateOne(
      {
        _id: new ObjectId(id),
      },
      {
        $set: article,
      },
    );

    debug({ mongoResult }, "mongoResult");

    if (!mongoResult.matchedCount) {
      throw new NotFoundError("No article found", { id });
    }

    return {
      id,
    };
  }
}

let repositoryStorage = new WeakMap<User, ArticlesRepository>();
export function createRepository(user: User): ArticlesRepository {
  let articlesRepository = repositoryStorage.get(user);
  if (articlesRepository) {
    return articlesRepository;
  }
  articlesRepository = new ArticlesRepository(user);
  repositoryStorage = new WeakMap();
  repositoryStorage.set(user, articlesRepository);
  return articlesRepository;
}
