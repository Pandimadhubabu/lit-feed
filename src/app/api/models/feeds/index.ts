import { NotFoundError } from "@/app/api/errors";
import * as logger from "@/app/api/logger";
import { Feed } from "@/types";
import { ObjectId } from "mongodb";
import { Repository } from "../Repository";
import { feeds, mongoToObject } from "../mongo";
export class Feeds extends Repository<Feed> {
  async create({ body }: { body: Feed }): Promise<Feed> {
    const feed = Object.assign(
      {},
      {
        ...body,
        userId: this.user.id,
        updatedAt: new Date(),
      },
    );
    const feedsCollection = await feeds();

    const { insertedId } = await feedsCollection.insertOne(feed);

    logger.debug({ insertedId }, "mongoResult");

    const feedResult = mongoToObject<Feed>({
      ...feed,
      _id: insertedId,
    });

    logger.debug({ feedResult }, "addFeed result");

    return feedResult;
  }

  async get({
    params: { feedId },
  }: {
    params: { feedId: Feed["id"] };
  }): Promise<Feed> {
    const feedsCollection = await feeds();

    const mongoResult = await feedsCollection.findOne({
      userId: this.user.id,
      _id: new ObjectId(feedId),
    });

    logger.debug({ mongoResult }, "mongoResult");

    if (!mongoResult) {
      throw new NotFoundError("No feed found", { feedId });
    }

    const feedResult = mongoToObject<Feed>(mongoResult);

    logger.debug({ feedResult }, "getFeed result");

    return feedResult;
  }

  async getAll(): Promise<Feed[]> {
    const feedsCollection = await feeds();

    const mongoResult = await feedsCollection
      .find({
        userId: this.user.id,
      })
      .toArray();

    logger.debug({ mongoResult }, "mongoResult");

    const feedsList = mongoResult.map(mongoToObject<Feed>);

    logger.debug({ feedsList }, "feedsList");

    return feedsList;
  }

  async delete({ params: { feedId } }: { params: { feedId: Feed["id"] } }) {
    const feedsCollection = await feeds();

    const mongoResult = await feedsCollection.deleteOne({
      userId: this.user.id,
      _id: new ObjectId(feedId),
    });

    logger.debug({ mongoResult }, "mongoResult");

    if (!mongoResult.deletedCount) {
      throw new NotFoundError("No feed found to delete", { feedId });
    }

    return {
      id: feedId,
    };
  }

  async update({
    params: { feedId },
    body,
  }: {
    params: { feedId: Feed["id"] };
    body: Feed;
  }) {
    const feed = Object.assign({}, body);
    const feedsCollection = await feeds();

    const mongoResult = await feedsCollection.updateOne(
      {
        userId: this.user.id,
        _id: new ObjectId(feedId),
      },
      {
        $set: {
          ...feed,
          updatedAt: new Date(),
        },
      },
    );

    logger.debug({ mongoResult }, "mongoResult");

    if (!mongoResult.matchedCount) {
      throw new NotFoundError("No feed found to update", { feedId });
    }

    return {
      id: feedId,
    };
  }
}
