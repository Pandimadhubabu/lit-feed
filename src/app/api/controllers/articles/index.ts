import { Article, Feed, User } from "@/types";
import * as logger from "../../logger";
import { Repository, createRepository } from "../../models/Repository";
import { Articles } from "../../models/articles";
import { Feeds } from "../../models/feeds";
import { XMLParser } from "fast-xml-parser";

export class ArticlesController extends Repository<Article> {
  async execute({
    params: { feedId },
  }: {
    params: { feedId: Feed["id"] };
  }): Promise<void> {
    const feedsRepository = await createRepository(Feeds, this.user);
    const feed = await feedsRepository.get({ params: { feedId } });

    const articlesRepository = await createRepository(Articles, this.user);

    const oldArticles = await articlesRepository.getAll({ params: { feedId } });

    const partialArticles = await getPartialArticles(feed.href);

    const newArticles = partialArticles.filter(
      (partialArticle) =>
        !oldArticles.find(
          (oldArticle) => oldArticle.href === partialArticle.href,
        ),
    );

    if (newArticles.length === 0) {
      logger.debug({ feed }, "No new articles");
      return;
    }

    const articles = await enrichPartialArticles({
      partialArticles: newArticles,
      feed,
    });

    const result: Article[] = [];
    for (const article of articles) {
      article.feedId = feed.id;

      const savedArticle = await articlesRepository.create({
        params: { feedId: feed.id },
        body: article,
      });

      result.push(savedArticle);
    }

    logger.debug({ result }, "Refreshed articles");

    await feedsRepository.update({
      params: { feedId },
      body: {
        ...feed,
        updatedAt: new Date(),
      },
    });

    logger.debug({ feed }, "Updated feed");
  }
}

function enrichPartialArticles({
  partialArticles,
  feed,
}: {
  partialArticles: PartialArticle[];
  feed: Feed;
}): Article[] {
  const articles: Article[] = [];
  for (const partialArticle of partialArticles) {
    const article: Article = {
      ...partialArticle,
      feedId: feed.id,
      isRead: false,
      isSaved: false,
      feedName: feed.name,
      id: "",
    };

    articles.push(article);
  }

  logger.debug({ articles }, "Enriched articles");
  return articles;
}

async function getPartialArticles(url: string): Promise<PartialArticle[]> {
  logger.debug({ url }, "Fetching articles from feed");
  const rss = await fetch(url);

  const rssText = await rss.text();
  logger.debug({ rssText }, "Fetched articles from feed");

  return rssToArticles(rssText);
}

export function rssToArticles(rss: string): PartialArticle[] {
  const parser = new XMLParser();
  const jsonObj = parser.parse(rss);

  const items = jsonObj.rss.channel.item;

  const articles: PartialArticle[] = [];

  for (const item of items) {
    const article: PartialArticle = {
      title: item.title,
      summary: item.description,
      href: item.link,
      date: item.pubDate,
      image: item.image,
    };
    articles.push(article);
  }

  logger.debug({ articles }, "Parsed articles from feed");

  return articles;
}

type PartialArticle = Partial<Article> & {
  title: string;
  summary: string;
  href: string;
  date: string;
  image?: string;
};
