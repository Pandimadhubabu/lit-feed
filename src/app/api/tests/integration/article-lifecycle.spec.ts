import { Feed, Article } from "@/types";
import { testArticles, testFeeds, testUser } from "@/app/api/models/testData";
import { omit } from "../../utils";
import { createHandler } from "../../createHandler";
import { Articles } from "../../models/articles";
import { Feeds } from "../../models/feeds";

const user = testUser;
const getArticles = createHandler(Articles, "getAll");
const getArticle = createHandler(Articles, "get");
const deleteArticle = createHandler(Articles, "delete");
const createArticle = createHandler(Articles, "create");
const getFeeds = createHandler(Feeds, "getAll");
const deleteFeed = createHandler(Feeds, "delete");
const addFeed = createHandler(Feeds, "create");

describe("article lifecycle tests", () => {
  let feedId: Feed["id"];
  beforeAll(async () => {
    // Get all feeds
    const { data: allFeeds } = await getFeeds({
      user,
    });

    // Get all articles from all feeds
    const allArticles = (
      await Promise.all(
        allFeeds.map(async (feed: Feed) => {
          const { data: articles } = await getArticles({
            params: {
              feedId: feed.id,
            },
            user,
          });

          return articles;
        }),
      )
    ).flat();

    // Delete all articles
    await Promise.all(
      allArticles.map(async (article: Article) => {
        const { data, message } = await deleteArticle({
          params: {
            articleId: article.id,
          },
          user,
        });

        expect(message).toMatchSnapshot();
        expect(data.id).toBe(article.id);
      }),
    );

    // Delete all feeds
    await Promise.all(
      allFeeds.map(async (feed: Feed) => {
        const { data, message } = await deleteFeed({
          params: {
            feedId: feed.id,
          },
          user,
        });

        expect(message).toMatchSnapshot();
        expect(data.id).toBe(feed.id);
      }),
    );

    // Create feed
    const { data, message } = await addFeed({
      body: {
        ...testFeeds[0],
      },
      user,
    });
    feedId = data.id;
  });

  test("should be able to add a new article", async () => {
    const { data, message } = await createArticle({
      body: {
        ...testArticles[0],
        feedId,
      },
      params: {
        feedId,
      },
      user,
    });

    expect(message).toMatchSnapshot();
    expect(omit(data, ["id", "feedId"])).toEqual(
      omit(testArticles[0], ["id", "feedId"]),
    );
    testArticles[0].id = data.id;
    testArticles[0].feedId = data.feedId;
  });

  test("should be able to add a second article", async () => {
    const { data, message } = await createArticle({
      body: {
        ...testArticles[1],
        feedId,
      },
      params: {
        feedId,
      },
      user,
    });

    expect(message).toMatchSnapshot();
    expect(omit(data, ["id", "feedId"])).toEqual(
      omit(testArticles[1], ["id", "feedId"]),
    );
    testArticles[1].id = data.id;
    testArticles[1].feedId = data.feedId;
  });

  test("should be able to get all articles", async () => {
    const { data, message } = await getArticles({
      params: {
        feedId,
      },
      user,
    });

    expect(message).toMatchSnapshot();
    expect(data).toEqual(
      testArticles.sort(
        (a, b) => new Date(b.date!).getTime() - new Date(a.date!).getTime(),
      ),
    );
  });

  test("should be able to delete an article", async () => {
    const { data, message } = await deleteArticle({
      params: {
        articleId: testArticles[0].id,
      },
      user,
    });

    expect(message).toMatchSnapshot();
    expect(data.id).toBe(testArticles[0].id);
  });

  test("should be able to get all articles after delete", async () => {
    const { data, message } = await getArticles({
      params: {
        feedId,
      },
      user,
    });

    expect(message).toMatchSnapshot();
    expect(data).toEqual([testArticles[1]]);
  });

  test("should not be able to get a deleted article", async () => {
    expect(
      getArticle({
        params: {
          articleId: testArticles[0].id,
        },
        user,
      }),
    ).rejects.toThrowErrorMatchingSnapshot();
  });

  test("should not be able to update a deleted article", async () => {
    expect(
      deleteArticle({
        params: {
          articleId: testArticles[0].id,
        },
        user,
      }),
    ).rejects.toThrowErrorMatchingSnapshot();
  });

  test("should be able to delete a second article", async () => {
    const { data, message } = await deleteArticle({
      params: {
        articleId: testArticles[1].id,
      },
      user,
    });

    expect(message).toMatchSnapshot();
    expect(data.id).toBe(testArticles[1].id);
  });

  test("should be able to get all articles after delete", async () => {
    const { data, message } = await getArticles({
      params: {
        feedId,
      },
      user,
    });

    expect(message).toMatchSnapshot();
    expect(data).toEqual([]);
  });
});
