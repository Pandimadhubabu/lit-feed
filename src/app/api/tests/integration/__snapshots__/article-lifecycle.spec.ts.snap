// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`article lifecycle tests should be able to delete a second article 1`] = `"No article found"`;

exports[`article lifecycle tests should be able to get all articles after delete 1`] = `"No article found"`;
